src/main/java/com/google/sps/servlets/GoServlet.java

```diff
@@ -0,0 +1,301 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Serves key information from optimizing between addresses. */
+@WebServlet("/go")
+public class GoServlet extends AuthenticatedHttpServlet {
+
+  private final DirectionsClientFactory directionsClientFactory;
+  private final PlacesClientFactory placesClientFactory;
+  private final TasksClientFactory tasksClientFactory;
+  private final GeocodingClientFactory geocodingClientFactory;
+  private final String apiKey;
+
+  /**
+   * Construct servlet with default DirectionsClient.
+   *
+   * @throws IOException
+   */
+  public GoServlet() throws IOException {
+    directionsClientFactory = new DirectionsClientImpl.Factory();
+    placesClientFactory = new PlacesClientImpl.Factory();
+    tasksClientFactory = new TasksClientImpl.Factory();
+    geocodingClientFactory = new GeocodingClientImpl.Factory();
+    apiKey = (new KeyProvider()).getKey("apiKey");
+  }
+
+  /**
+   * Construct servlet with explicit implementation of DirectionsClient.
+   *
+   * @param factory A DirectionsClientFactory containing the implementation of
+   *     DirectionsClientFactory.
+   */
+  public GoServlet(
+      DirectionsClientFactory directionsClientFactory,
+      PlacesClientFactory placesClientFactory,
+      TasksClientFactory tasksClientFactory,
+      GeocodingClientFactory geocodingClientFactory,
+      String apiKey) {
+    this.directionsClientFactory = directionsClientFactory;
+    this.placesClientFactory = placesClientFactory;
+    this.tasksClientFactory = tasksClientFactory;
+    this.geocodingClientFactory = geocodingClientFactory;
+    this.apiKey = apiKey;
+  }
+
+  /**
+   * Returns the most optimal order of travel between addresses.
+   *
+   * @param request HTTP request from the client.
+   * @param response HTTP response to the client.
+   * @throws ServletException
+   * @throws IOException
+   */
+  @Override
+  public void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws ServletException, IOException {
+    assert googleCredential != null
+        : "Null credentials (i.e. unauthenticated requests) should already be handled";
+    // Get all tasks from user's tasks account TODO: Get relevant tasks using task
+    // titles
+    TasksClient tasksClient = tasksClientFactory.getTasksClient(googleCredential);
+    DirectionsClient directionsClient = directionsClientFactory.getDirectionsClient(apiKey);
+
+    // Initialize Tasks Response
+    List<Task> tasks;
+
+    String taskLists = request.getParameter("taskLists");
+    if (taskLists == null) {
+      tasks = TasksServlet.getAllTasksFromAllTaskLists(tasksClient);
```

Comment:
Initially yes, but not anymore because these methods have been moved to a utility class in https://github.com/googleinterns/blueprint-step-2020/pull/146.

Reply:
`RESOLVED`